App Manager V2 Available Under Feature Flag

biyeun opened the pull request:
Here it is! The Big PR that makes App Manager V2 available under a feature flag. (finally)
The sooner this is reviewed and merged, the easier all the conflicts can be handled, and people can view changes side by side in both versions.

I did a local QA of V1, and everything functions as expected. All the template and style changes are by default not user facing. Sanity check the `*.py` files, but those should all be really simple or very clearly isolated/flagged changes.

@orangejenny / @benrudolph / @proteusvacuum 
fyi @calellowitz because you have a big PR out

One step closer to a new HQ!

biyeun committed:
update partial template refs

biyeun committed:
update template refs to be toggled

biyeun committed:
update js to match original branch for onboarding

biyeun committed:
add templatetag for prefixing translation

biyeun committed:
handle split default / app settings views for v2

biyeun committed:
add option for releases context

biyeun committed:
partial build errors template support

biyeun committed:
updates for release manager solo page

biyeun committed:
copy update

biyeun committed:
redirect to view_app for app manager v2

biyeun committed:
v2 template

biyeun committed:
copy update

biyeun committed:
serve v2 template

biyeun committed:
update css interactions

biyeun committed:
add inline edit v2 component

biyeun committed:
update styles

biyeun committed:
update look and feel of app settings

biyeun committed:
add inline_edit_trans_v2 tag

biyeun committed:
add toggle to simplify parts of hq for user testing

biyeun committed:
wrap save button in savebtn-bar

biyeun committed:
remove navigation for user testing toggle

biyeun committed:
hack fix to column layout

biyeun committed:
update inline_edit_v2

biyeun committed:
redirect to apps dashboard on app deletion

biyeun committed:
HIDE edit form notice on app-v2 views only

biyeun committed:
style updates

biyeun committed:
update titles and reorder tabs

biyeun committed:
add isVisible to app settings

biyeun committed:
extend from two column we don't need anything from section

biyeun committed:
update button options in form designer view

biyeun committed:
fix form tab advance style and modal issues

biyeun committed:
update form workflow panels

biyeun committed:
don't allow user case for user testing this time

biyeun committed:
make sure a default template is set here

biyeun committed:
update usercase_config panels

biyeun committed:
update panels

biyeun committed:
update intro text and titles

biyeun committed:
fix app name update

biyeun committed:
update module view titles

biyeun committed:
hide set path for user testing

biyeun committed:
update panels

biyeun committed:
update case config views

biyeun committed:
release manager styling

biyeun committed:
fix popover issues nested within hq-content

biyeun committed:
cleanup view urls in app manager

biyeun committed:
remove non existing view

biyeun committed:
update urls

biyeun committed:
fix lingering template syntax errors from bad merges

biyeun committed:
update old url

biyeun committed:
update look and feel of module settings page

biyeun committed:
update form settings templates

biyeun committed:
update url

biyeun committed:
update diffs

dimagimon commented:
There are 8 errors:

* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 210 - 'data' is defined but never used (no-unused-vars)
* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 302 - Missing trailing comma. (comma-dangle)
* corehq/apps/app_manager/static/app_manager/js/app_manager_media.js, line 155 - 'COMMCAREHQ' is not defined. (no-undef)
* corehq/apps/style/static/style/js/main.js, line 144 - 'k' is defined but never used (no-unused-vars)
* corehq/apps/style/static/style/ko/components.ko.js, line 4 - Missing trailing comma. (comma-dangle)
* corehq/apps/app_manager/templatetags/xforms_extras.py, line 46 - E302 expected 2 blank lines, found 1
* corehq/apps/app_manager/views/modules.py, line 837 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views/modules.py, line 840 - E121 continuation line under-indented for hanging indent

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -247,12 +282,50 @@ hqDefine('app_manager/js/app_manager.js', function () {
+        if (COMMCAREHQ.toggleEnabled('APP_MANAGER_V2')) {
+
+            $('#js-add-new-item').popover({
+                title: "Add",
```
translate

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -247,12 +282,50 @@ hqDefine('app_manager/js/app_manager.js', function () {
+                $(this).popover('show');
+            });
+
+            $('body').click(function (event) {
```
This logic seems pretty similar to using `focus` for the popover trigger, curious if that would have worked instead.

orangejenny commented on corehq/apps/app_manager/templates/app_manager/v2/app_view.html:
```diff
@@ -61,39 +61,50 @@
-    </a>
+{% block pre_page_content %}
+
+  <div class="appmanger-page-actions">
```
fix spelling (looks like this is spelled consistently within the PR, so not a bug, just for the sake of sanity)

orangejenny commented on corehq/apps/app_manager/templates/app_manager/v2/form_designer.html:
```diff
@@ -158,16 +158,55 @@
 
         $(function () {
+
+        });
```
looks like this & the above line can be deleted

orangejenny commented on corehq/apps/app_manager/templates/app_manager/v2/app_view.html:
```diff
@@ -61,39 +61,50 @@
+
+  <div class="appmanger-page-actions">
+      <a class="btn btn-purple js-preview-toggle" style="display: none;" href="#">
+        <span class="js-preview-action-show hide"><i class="fa fa-play-circle"></i> {% trans "Preview App" %}</span>
```
Is the "Preview App" text intentionally the same for both actions?

biyeun replied:
yep

orangejenny commented on corehq/apps/app_manager/templates/app_manager/v2/partials/case_config.html:
```diff
@@ -28,50 +28,54 @@ <h4 class="modal-title">{% trans "Remove Subcase?" %}</h4>
     <div data-bind="saveButton: saveButton"></div>
+
+    <p class="lead">
+        {% if not app.advanced_app_builder %}
```
translations

orangejenny commented on corehq/apps/app_manager/templates/app_manager/v2/partials/case_config_shared.html:
```diff
@@ -64,7 +68,18 @@
             valueDefault: property.key,
             default: property.defaultKey,
-            casePropertySelect2: suggestedProperties,
+            casePropertyTypeahead: suggestedProperties,
```
Possible bad merge? `casePropertyTypeahead` was removed in https://github.com/dimagi/commcare-hq/pull/13807

biyeun replied:
thanks! changing....definitely a bad merge

orangejenny commented on corehq/apps/app_manager/templates/app_manager/v2/partials/commcare_settings.html:
```diff
@@ -3,91 +3,97 @@
+      <div class="panel-collapse collapse" data-bind="attr: {id: id}, css: {'in': !reallyCollapse()}">
+        <div class="panel-body">
+          <!-- ko foreach: settings -->
+          <!-- ko if: widget !== 'none' -->
```
Possible bad merge - the `none` widget was removed in https://github.com/dimagi/commcare-hq/pull/13963

biyeun replied:
thanks, updating

orangejenny commented on corehq/apps/app_manager/templates/app_manager/v2/partials/releases.html:
```diff
@@ -70,7 +70,7 @@
 
 {% if intro_only %}
-    Intro content goes here. Why don't you click on <b>New</b> on the left?
+    Please click <strong>Add...</strong> to the left to start building your application!
```
trans

biyeun replied:
done

orangejenny commented on corehq/apps/app_manager/templatetags/xforms_extras.py:
```diff
@@ -17,29 +17,35 @@ def translate(t, lang, langs=None):
 @register.filter
 def html_trans(name, langs=["default"]):
-    return mark_safe(html.strip_tags(trans(name, langs, use_delim=False)) or EMPTY_LABEL)
+    return mark_safe(trans(name, langs, use_delim=False) or EMPTY_LABEL)
```
Why is `strip_tags` no longer needed?

biyeun replied:
`strip_tags` might have been added after I made this change, and so this was a bad merge

biyeun replied:
nope, actually that is intentional. I give the lang prefix a class so that looks like 
<img width="151" alt="screen shot 2016-12-22 at 4 06 26 pm" src="https://cloud.githubusercontent.com/assets/716573/21440372/b385b262-c860-11e6-967e-82a444bf8c39.png">

If I add the strip_tags, it removes that html

orangejenny replied:
Removing `strip_tags` opens us up to javascript injection. Could fall back to doing text prefixes temporarily (`[en] Survey`) rather than deal with finding an alternative approach right now.

orangejenny commented on corehq/apps/style/static/style/ko/components/inline_edit_v2.js:
```diff
@@ -0,0 +1,182 @@
+/* global DOMPurify */
```
Wondering if it's be worth moving this file into app manager temporarily, for the sake of taking advantage of the v1/v2 split and the diffs. Thoughts?

orangejenny commented on corehq/toggles.py:
```diff
@@ -934,7 +934,7 @@ def _commtrackify(domain_name, toggle_is_enabled):
     'app_manager_v2',
     'Prototype for case management onboarding (App Manager V2)',
-    TAG_EXPERIMENTAL,
+    TAG_PRODUCT_PATH,
```
:)

orangejenny commented:
This looks great. The only potentially blocking comments are
- the two places that look like possible bad merges
- understanding why the javascript injection code (`strip_tags`) was removed

Do you want this merged & potentially deployed? Guessing yes, just wanted to confirm.

biyeun committed:
translations

biyeun committed:
biyeun can't spell

biyeun committed:
cleanup js

biyeun committed:
update bad merges

biyeun committed:
add back strip tags for regular html trans

biyeun committed:
add style for header langcode

dimagimon commented:
There are 9 errors:

* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 210 - 'data' is defined but never used (no-unused-vars)
* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 294 - 'django' is not defined. (no-undef)
* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 302 - Missing trailing comma. (comma-dangle)
* corehq/apps/app_manager/static/app_manager/js/app_manager_media.js, line 155 - 'COMMCAREHQ' is not defined. (no-undef)
* corehq/apps/style/static/style/js/main.js, line 144 - 'k' is defined but never used (no-unused-vars)
* corehq/apps/style/static/style/ko/components.ko.js, line 4 - Missing trailing comma. (comma-dangle)
* corehq/apps/app_manager/templatetags/xforms_extras.py, line 46 - E302 expected 2 blank lines, found 1
* corehq/apps/app_manager/views/modules.py, line 837 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views/modules.py, line 840 - E121 continuation line under-indented for hanging indent

biyeun commented:
@orangejenny addressed most of the concerns inline

most importantly i fixed the bad merges, thanks for catching those.
I also added the strip_tags back to the regular html trans since that's being used on prod, but I still need the one for the prefix as that's how I style the translation tag in the app manager navigation. This new template tag is only being used in v2 right now.

confirming that, yes, I would like to get this merged and deployed asap. :)

orangejenny commented:
In `case_config_shared.html`, the PR that changed `casePropertyTypeahead` to `casePropertySelect2` also removed the `event` logic in that input. Once that's removed and tests pass, feel free to self-merge. Excited to see this moving forward!

biyeun committed:
also remove event logic

biyeun committed:
update diffs

dimagimon commented:
There are 9 errors:

* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 210 - 'data' is defined but never used (no-unused-vars)
* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 294 - 'django' is not defined. (no-undef)
* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 302 - Missing trailing comma. (comma-dangle)
* corehq/apps/app_manager/static/app_manager/js/app_manager_media.js, line 155 - 'COMMCAREHQ' is not defined. (no-undef)
* corehq/apps/style/static/style/js/main.js, line 144 - 'k' is defined but never used (no-unused-vars)
* corehq/apps/style/static/style/ko/components.ko.js, line 4 - Missing trailing comma. (comma-dangle)
* corehq/apps/app_manager/templatetags/xforms_extras.py, line 46 - E302 expected 2 blank lines, found 1
* corehq/apps/app_manager/views/modules.py, line 837 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views/modules.py, line 840 - E121 continuation line under-indented for hanging indent

biyeun commented:
@orangejenny  updated

biyeun commented:
@czue 

fyi when tests pass on this the plan is to have it merged into master to avoid further conflicts. All changes are behind a feature flag, but if you do notice anything wonky in app builder (which shouldn't happen) this might be the first place to check. I'll be online the next two weeks anyway. :)

czue commented:
thanks for the heads up and congrats on the milestone!

czue commented:
fyi, I will likely plan to revert this if there are interrupt level issues and you're not around to address. if you prefer not to take that risk then it might be easier to merge on Jan 1, though totally up to you.

benrudolph reviewed (approved):
🚀 went through it the best i could, trusting your local testing. great job on getting this this far!

dimagimon commented:
There are 9 errors:

* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 210 - 'data' is defined but never used (no-unused-vars)
* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 294 - 'django' is not defined. (no-undef)
* corehq/apps/app_manager/static/app_manager/js/app_manager.js, line 302 - Missing trailing comma. (comma-dangle)
* corehq/apps/app_manager/static/app_manager/js/app_manager_media.js, line 155 - 'COMMCAREHQ' is not defined. (no-undef)
* corehq/apps/style/static/style/js/main.js, line 144 - 'k' is defined but never used (no-unused-vars)
* corehq/apps/style/static/style/ko/components.ko.js, line 4 - Missing trailing comma. (comma-dangle)
* corehq/apps/app_manager/templatetags/xforms_extras.py, line 46 - E302 expected 2 blank lines, found 1
* corehq/apps/app_manager/views/modules.py, line 864 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views/modules.py, line 867 - E121 continuation line under-indented for hanging indent

dannyroberts commented on corehq/apps/app_manager/static/app_manager/js/case-config-ui-advanced.js:
```diff
@@ -402,7 +402,7 @@ hqDefine('app_manager/js/case-config-ui-advanced.js', function () {
                         fixture_variable: '',
                         case_property: '',
-                        auto_select: false,
+                        auto_select: (COMMCAREHQ.toggleEnabled('APP_MANAGER_V2')) ? '' : false,
```
@biyeun looks like this line from this PR 7 months ago is causing https://manage.dimagi.com/default.asp?257934. Essentially, if you add a "Load Case From Fixture" Action to an Advanced Form, and then save immediately, the server 500s because it can't wrap `auto_select` value '' since it's expecting a boolean.

If I just undo this line and change it back to `false` it seems like everything works. Any idea if I'm missing something in the intention behind this line?
